debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.relational;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.jdbc.JdbcConnection;
+import io.debezium.pipeline.source.spi.SnapshotChangeEventSource;
+import io.debezium.pipeline.source.spi.StreamingChangeEventSource;
+import io.debezium.pipeline.spi.OffsetContext;
+import io.debezium.pipeline.spi.SnapshotResult;
+import io.debezium.schema.SchemaChangeEvent;
+
+/**
+ * Base class for {@link SnapshotChangeEventSource} for relational databases with a schema history.
+ *
+ * @author Gunnar Morling
+ */
+// TODO Mostly, this should be usable for Postgres as well; only the aspect of managing the schema history will have to
+// be made optional based on the connector
+public abstract class HistorizedRelationalSnapshotChangeEventSource implements SnapshotChangeEventSource {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HistorizedRelationalSnapshotChangeEventSource.class);
+
+    private final RelationalDatabaseConnectorConfig connectorConfig;
+    private final OffsetContext previousOffset;
+    private final JdbcConnection jdbcConnection;
+    private final HistorizedRelationalDatabaseSchema schema;
+
+    public HistorizedRelationalSnapshotChangeEventSource(RelationalDatabaseConnectorConfig connectorConfig, OffsetContext previousOffset, JdbcConnection jdbcConnection, HistorizedRelationalDatabaseSchema schema) {
+        this.connectorConfig = connectorConfig;
+        this.previousOffset = previousOffset;
+        this.jdbcConnection = jdbcConnection;
+        this.schema = schema;
+    }
+
+    @Override
+    public SnapshotResult execute(ChangeEventSourceContext context) throws InterruptedException {
+        // for now, just simple schema snapshotting is supported which just needs to be done once
+        if (previousOffset != null) {
+            LOGGER.debug("Found previous offset, skipping snapshotting");
+            return SnapshotResult.completed(previousOffset);
+        }
+
+        Connection connection = null;
+
+        try (SnapshotContext ctx = prepare(context)) {
+            LOGGER.info("Snapshot step 1 - Preparing");
+
+            connection = jdbcConnection.connection();
+            connection.setAutoCommit(false);
+
```

Comment:
A lockDatabase step should be added - for the databases that support it.

Reply:
You mean distinct from the `lockTables()` below? I'd prefer to add this once there's a DB and implementation where we need it. Otherwise it's hard to do it correctly now, not knowing the specific needs. This base class is definitely not cast in stone, it will have to be adjusted as we add more implementations to it.